lib/Transforms/ConvertToCiphertextSemantics/ConvertToCiphertextSemantics.cpp

```diff
@@ -228,54 +239,241 @@ class ConvertAssignLayout
   using ContextAwareOpConversionPattern<
       tensor_ext::AssignLayoutOp>::ContextAwareOpConversionPattern;
 
+  Value expandDims(Value value, LayoutAttr layout,
+                   ImplicitLocOpBuilder &b) const {
+    RankedTensorType dataSemanticType = cast<RankedTensorType>(value.getType());
+    tensor_ext::AlignmentAttr alignment = layout.getAlignment();
+
+    // It's a bit weird, but to make an expand shape op we have to group the
+    // output indices in dataSemanticType.getRank() many groups where the 1's
+    // are all grouped with axes from the dataSemanticType. But the 1's can
+    // show up before or after the data semantic tensor's dims, so we
+    // eagerly consume unit dims before and after each data semantic dim.
+    SmallVector<int64_t> newSizes;
+    SmallVector<ReassociationIndices> reassociation;
+    ReassociationIndices nextGroup;
+    int64_t ciphertextIndex = 0, groupIndex = 0;
+    while (groupIndex < dataSemanticType.getRank()) {
+      // Process all the unit dims.
+      while (containsDim(alignment.getInsertedDims(), ciphertextIndex)) {
+        newSizes.push_back(1);
+        nextGroup.push_back(ciphertextIndex);
+        ++ciphertextIndex;
+      }
+
+      // Now process exactly one data dim.
+      newSizes.push_back(dataSemanticType.getDimSize(groupIndex));
+      nextGroup.push_back(ciphertextIndex);
+      ++ciphertextIndex;
+
+      // Now process any more unit dims.
+      while (containsDim(alignment.getInsertedDims(), ciphertextIndex)) {
+        newSizes.push_back(1);
+        nextGroup.push_back(ciphertextIndex);
+        ++ciphertextIndex;
+      }
+
+      reassociation.push_back(nextGroup);
+      nextGroup.clear();
+      ++groupIndex;
+    }
+    RankedTensorType expandedType =
+        RankedTensorType::get(newSizes, dataSemanticType.getElementType());
+    auto expandOp =
+        b.create<tensor::ExpandShapeOp>(expandedType, value, reassociation);
+    setMaterializedAttr(expandOp);
+    expandOp->setAttr(kLayoutAttrName, layout);
+    return expandOp.getResult();
+  }
+
+  Value applyPadding(Value value, LayoutAttr layout,
+                     ImplicitLocOpBuilder &b) const {
+    RankedTensorType dataSemanticType = cast<RankedTensorType>(value.getType());
+    tensor_ext::AlignmentAttr alignment = layout.getAlignment();
+    auto padValueOp = b.create<arith::ConstantOp>(alignment.getPaddingValue());
```

Comment:
padding value is a default nullptr - (and maybe will also be unspecified later) so it probably should check and return early in those cases?

edit: oh i see this is checked before the call to this func

Reply:
Added a comment in 80decf67